File: Search/NextPermutation/NextPermutations.py
class Solution:
    def nextPermutation(self, num):
        if num is None or len(num) <= 1:
            return None
            
        i = 0
        for i in xrange(len(num) - 1, -1, -1):
            if num[i] < num[i + 1]:
                break
            elif i == 0:
                num = num[::-1]
                return num
                
        j = 0
        for j in xrange(len(num) - 1, i, -1):
            if num[i] < num[j]:
                break
                
        num[i], num[j] = num[j], num[i]
        num[i + 1: len(num)] = num[len(num) - 1:i:-1]
            
        return num
        
                
                
                

File: Search/NextPermutation/NextPermutations.java
class Solution {
    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length == 0)
            return;
        
        int k = -1;
        for(int i = nums.length - 1; i >= 0; i--) {
            if(nums[i] < nums[i + 1])
                k = i;
                break;            
        }
        
        if(k == -1) {
            reverse(nums, 0, nums.length - 1);
            return;
            
        }
        
        int l = nums.length - 1;
        while(l > k && nums[l] <= nums[k]) l--;
        
        int tmp = nums[k];
        nums[k] = nums[l];
        nums[l] = tmp;
        
        reverse(nums, k + 1, nums.length - 1);
    }
    
    private void reverse(int[] nums, int start, int end) {
        for(int i = start, j = end; i < j; i++, j--) {
            int tmp = nums[i];
            nums[i] = nums[j];
            nums[j] = tmp;
            
        }
    }
}